### PR TITLE
:racehorse: Add feature [Maintenance Mode]

### PR DESCRIPTION
**Pull request by londry**
The system maintenance mode helps you improve the user experience.
This function can be turned on in the `bb-config.php` file.
Just set
```
'maintenance_mode' => true,
```
to enabled maintenance mode. Once turned on, your users will
receive a maintenance notification when accessing the API and front store.
This process will not interfere with the administrator login and use of the
admin area interface.

For redesign your maintenance page, you can edit template file at:
`/bb-themes/boxbilling/html/maintenance.phtml`

Issues: https://github.com/boxbilling/boxbilling/issues/1094

### CODE REVIEW

**Comment by BenNottelling** [comment]
> Both config file and admin area at the same time is okay? But there will have a priority to ensure that it can still be works fine when the administrator interface is not accessible.
> 
> * bb-config.php [High Priority]
> * Admin area [Base Priority]

Why would there be a priority?

Just make it if it's enabled in either location that it'll display the maintenance mode screen

**Comment by londry** [comment]
I want to give up making this decision. Therefore I am willing to maintain the use of `bb-config.php` to manage the maintenance mode. At least, who should be the webmaster or technical engineer to enable this mode, because the operation of `bb-config.php` should not feel difficult.

**Comment by evrifaessa** [comment]
> I want to give up making this decision. Therefore I am willing to maintain the use of `bb-config.php` to manage the maintenance mode. At least, who should be the webmaster or technical engineer to enable this mode, because the operation of `bb-config.php` should not feel difficult.

Well, I think making the option configurable from the web interface can be a future task. But I'd be glad if someone gets tests working before merging this.

**Comment by londry** [comment]
> @londry I'm okay to have an option in the admin area. Also I think we should move the admin area paths to the config instead of hardcoding them.

Depending on the user's installation path, the Admin Path will change in the future. We can't guess whether the user will change these configurations if in `bb-config.php`. I think the best way is to hard code currentlly, because it can adapt to changes in the installation directory.

**Comment by londry** [comment]
If there are no other objections, we can facilitate the merger process to close this topic. Its time to make a decision. @evrifaessa

**Comment by evrifaessa** [comment]
> If there are no other objections, we can facilitate the merger process to close this topic. Its time to make a decision. @evrifaessa

Well, I already said the tests have to run before I can merge anything.

File: tests/bb-library/Box/Box_RequestTest.php [hunk 1]
```diff
@@ -295,6 +295,7 @@ public function testGetClientAddressForwarded()
 
         $this->assertEquals('123.123.123.120', $request->getClientAddress(true));
```
[the remark this answers is not in the document]
Comment (John-S4) [reply tests/bb-library/Box/Box_RequestTest.php:296]: Me three, I also can't see any obvious reasons, strange. Will take another look with fresh eyes tomorrow.

**Comment by BenNottelling** [comment]
> I implemented Cloudflare's CF-Connecting-IP, used Box_Request to get the visitor's IP and made a few other changes. It seems good to merge, but I still would like to hear some reviews from other maintainers.

Do we need these changes for this PR?
It's good feature, and those changes are what caused the failing tests

**Comment by evrifaessa** [comment]
> Do we need these changes for this PR? It's good feature, and those changes are what caused the failing tests

I've reverted what possibly could break the tests. Now it should be good to go. We can implement Cloudflare's headers later.

**Review by BenNottelling — APPROVED** [review]
I'm happy with this

**Pull request merged**